File: sim/src/core/transit/routePath.ts
/**
 * Derived route geometry: the out-and-back polyline a route's vehicles follow.
 * Cached per route; cache key covers the segment list so edits invalidate.
 */
import { makePolyline } from '../geometry';
import type { Polyline, Vec2 } from '../geometry';
import type { GameState, RouteDef } from '../types';

const cache = new Map<string, { key: string; path: Polyline }>();

export function clearRoutePathCache(): void {
  cache.clear();
}

export function getRoutePath(state: GameState, route: RouteDef): Polyline | null {
  // scope the memo to the game instance + route: entity ids reset per newGame, so a
  // bare route.id key would return another game's stale polyline when one process
  // serves multiple games (sidecar/replay).
  const cacheKey = `${state.instanceId}:${route.id}`;
  const key = route.segmentIds.join(',');
  const hit = cache.get(cacheKey);
  if (hit && hit.key === key) return hit.path;

  const outbound: Vec2[] = [];
  for (let i = 0; i < route.segmentIds.length; i++) {
    const seg = state.tracks.find((t) => t.id === route.segmentIds[i]);
    if (!seg) return null;
    const fromId = route.stationIds[i];
    let pts = seg.polyline.points;
    if (seg.toStationId === fromId) pts = [...pts].reverse();
    // skip duplicate joint point
    const startAt = outbound.length > 0 ? 1 : 0;
    for (let j = startAt; j < pts.length; j++) outbound.push(pts[j] as Vec2);
  }
  if (outbound.length < 2) return null;
  const back = [...outbound].reverse().slice(1);
  const path = makePolyline([...outbound, ...back]);
  cache.set(cacheKey, { key, path });
  return path;
}

File: crates/mf-sim/src/transit/route_path.rs
//! Derived route geometry: the out-and-back polyline a route's vehicles follow.
//! Port of `sim/src/core/transit/routePath.ts`.
//!
//! The TS side memoizes per `instanceId:routeId`; the Rust port computes fresh
//! (the callers are cold paths / lane-B movement setup, not a hashed loop).

use crate::geometry::{make_polyline, Polyline, Vec2};
use crate::types::{GameState, RouteDef};

/// Out-and-back polyline for a route, or `None` if a segment is missing or the
/// path degenerates. Mirrors `getRoutePath`.
pub fn get_route_path(state: &GameState, route: &RouteDef) -> Option<Polyline> {
    let mut outbound: Vec<Vec2> = Vec::new();
    for i in 0..route.segment_ids.len() {
        let seg = state.tracks.iter().find(|t| t.id == route.segment_ids[i])?;
        let from_id = route.station_ids[i];
        let mut pts: Vec<Vec2> = seg.polyline.points.clone();
        if seg.to_station_id == from_id {
            pts.reverse();
        }
        let start_at = if outbound.is_empty() { 0 } else { 1 };
        for p in pts.into_iter().skip(start_at) {
            outbound.push(p);
        }
    }
    if outbound.len() < 2 {
        return None;
    }
    let back: Vec<Vec2> = outbound.iter().rev().skip(1).copied().collect();
    let mut all = outbound;
    all.extend(back);
    Some(make_polyline(all))
}
